sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveExternalCatalog.scala

```diff
@@ -200,22 +348,73 @@ private[spark] class HiveExternalCatalog(client: HiveClient, hadoopConf: Configu
    * Alter a table whose name that matches the one specified in `tableDefinition`,
    * assuming the table exists.
    *
-   * Note: As of now, this only supports altering table properties, serde properties,
-   * and num buckets!
+   * Note: As of now, this only supports altering table properties and serde properties.
    */
   override def alterTable(tableDefinition: CatalogTable): Unit = withClient {
     assert(tableDefinition.identifier.database.isDefined)
     val db = tableDefinition.identifier.database.get
     requireTableExists(db, tableDefinition.identifier.table)
-    client.alterTable(tableDefinition)
+    verifyTableProperties(tableDefinition)
+
+    if (tableDefinition.provider == Some("hive") || tableDefinition.tableType == VIEW) {
+      client.alterTable(tableDefinition)
+    } else {
+      val oldDef = client.getTable(db, tableDefinition.identifier.table)
+      // Sets the `schema`, `partitionColumnNames` and `bucketSpec` from the old table definition,
+      // to retain the spark specific format if it is.
+      // Also add table meta properties to table properties, to retain the data source table format.
+      val newDef = tableDefinition.copy(
+        schema = oldDef.schema,
+        partitionColumnNames = oldDef.partitionColumnNames,
+        bucketSpec = oldDef.bucketSpec,
+        properties = tableMetadataToProperties(tableDefinition) ++ tableDefinition.properties)
```

Comment:
if tableMetadataToProperties(tableDefinition) has new schema, `tableMetadataToProperties(tableDefinition)` is actually changing the schema?